dotnetv3/EC2/VirtualPrivateCloudExamples/CreateVPCforS3Example/CreateVPCforS3.cs

```diff
@@ -0,0 +1,136 @@
+
+        var s3EndpointUrl = await CreateVPCforS3Client(configuration, ec2Client);
+
+        if (s3EndpointUrl != null)
+        {
+            var objectsList = await CreateS3ClientWithEndpoint(configuration, s3EndpointUrl);
+            Console.WriteLine("List of objects in bucket:");
+            Console.WriteLine(string.Join(" ", objectsList.Select(o => o.Key)));
+        }
+    }
+
+
+    // snippet-start:[EC2.dotnetv3.CreateVPCforS3]
+
+    /// <summary>
+    /// Create a VPC Endpoint and use it for an S3 client.
+    /// </summary>
+    /// <param name="configuration">Configuration to specify resource ids.</param>
+    /// <param name="ec2Client">Initialized EC2 client.</param>
+    /// <returns>The S3 url using the endpoint.</returns>
+    public static async Task<string?> CreateVPCforS3Client(IConfiguration configuration,
+        AmazonEC2Client ec2Client)
+    {
+        var endpointResponse = await ec2Client.CreateVpcEndpointAsync(
+            new CreateVpcEndpointRequest
+            {
+                VpcId = configuration["VpcId"],
+                VpcEndpointType = VpcEndpointType.Interface,
+                ServiceName = "com.amazonaws.us-east-1.s3",
+                SubnetIds = new List<string> { configuration["SubnetId"]! },
+                SecurityGroupIds = new List<string> { configuration["SecurityGroupId"]! },
+                TagSpecifications = new List<TagSpecification>
+                {
+                    new TagSpecification
+                    {
+                        ResourceType = ResourceType.VpcEndpoint,
+                        Tags = new List<Tag>
+                        {
+                            new Tag("service", "S3")
+                        }
+                    }
+                }
+            });
+
+        var newEndpoint = endpointResponse.VpcEndpoint;
+
+        Console.WriteLine(
+            $"VPC Endpoint {newEndpoint.VpcEndpointId} was created, waiting for it to be available. " +
+            $"This may take a few minutes.");
+
+        State? endpointState = null;
+
+        while (endpointState == null ||
+               !(endpointState == State.Available || endpointState == State.Failed))
+        {
+            var endpointDescription = await ec2Client.DescribeVpcEndpointsAsync(
+                new DescribeVpcEndpointsRequest
+                {
+                    VpcEndpointIds = new List<string>
+                        { endpointResponse.VpcEndpoint.VpcEndpointId }
+                });
+            endpointState = endpointDescription.VpcEndpoints.FirstOrDefault()?.State;
+            Thread.Sleep(500);
+        }
+
+        if (endpointState == State.Failed)
+        {
+            Console.WriteLine("VPC Endpoint failed.");
+            return null;
+        }
+
+        // For newly created endpoints we may need to wait a few
+        // more minutes before using it for the Amazon S3 client.
+        Thread.Sleep(int.Parse(configuration["WaitTimeMilliseconds"]!));
```

Comment:
Yes, a retry is another approach that could work. The failures were inconsistent when using even after the state was Available, so I added the optional wait as a fallback, mainly because I wouldn't expect the user to create then immediately use the endpoint IRL. I like the retry in javascript, I have done something like that in some cases in C# also. But it doesn't look very much like what you got 😁